### PR TITLE
fix: filter undone event groups before moving them to the redo stack

### LINKED ISSUE #7026: Merged events producing no-op event result in incorrect state after repeated undo/redo.
### Description

eventUtils.filter(...) attempts to merge related events into a single change for broadcast purposes, but in doing so it modifies the underlying events which remain in the undo/redo history. Those modified events can become no-op events even if they weren't before, reporting true for isNull(). The next time filtering is performed, these now no-op events are skipped by the merging process and important data is lost, resulting in an incorrect state in the workspace after performing undo + redo + undo.

### Reproduction steps

1. Open the [advanced playground](https://blockly-demo.appspot.com/static/tests/playgrounds/advanced_playground.html).
2. From the Math category, drag a new number value block into a workspace, with the default value "123".
3. Click the number to select it, and type "1", "2", "3", producing three block change events with the following data:
    * {oldValue: 123, newValue: 1},
    * {oldValue: 1, newValue: 12},
    * {oldValue: 12, newValue: 123},
5. Undo. The filtering and merging process overwrites the newValue of the first event to 123, and then the first event is run backward, using its oldValue 123 to update the field. The events now look like:
    * {oldValue: 123, newValue: 123}, (this is now a no-op!)
    * {oldValue: 1, newValue: 12},
    * {oldValue: 12, newValue: 123},
6. Redo. The first block is skipped entirely and the second block is now the basis for merging. It is merged with the third block and its newValue 123 is used to update the field.
    * {oldValue: 123, newValue: 123}, (still a no-op, ignored during filtering/merging)
    * {oldValue: 1, newValue: 123},
    * {oldValue: 12, newValue: 123},
7. Undo again. The first event is ignored entirely, so the second event's oldValue 1 is assigned to the field, resulting in an incorrect state for the field. The field did not originally have the value 1.

### PR DESCRIPTION
**Pull request by ewpatton**
## The basics

- [x] I branched from develop
- [x] My pull request is against develop
- [x] My code follows the [style guide](https://developers.google.com/blockly/guides/modify/web/style-guide)
- [x] I ran `npm run format` and `npm run lint`

### Resolves

Fixes #7026 

### Proposed Changes

This PR filters events in a group prior to undoing them.

#### Behavior Before Change

Events being undone were not filtered before being put onto the redo stack. If there were multiple events that needed to be merged, this can result in inconsistent changes on the undo and redo stacks.

#### Behavior After Change

During undo, events within a group are passed through the Event.utils.filter function so that any sequence of events that result in a no-op will be removed from the undo/redo stack.

### Reason for Changes

The issue was appropriately described in #7026.

### Test Coverage

I ran `npm test` and all tests passed. However, we may want to add more tests to evaluate the specific sequence of events that lead to #7026.

### CODE REVIEW

**Review by BeksOmega — APPROVED** [review]
This LGTM :D

My plan is just to accept this as is since it fixes the bug, but I think John's other idea is also a good one:

> I would also suggest testing to see if the output of filtering/merging contains any non-null events. If not, then the undo/redo behavior should be repeated, trying the next group of events, until something actually changes, otherwise it feels weird to press undo and nothing happens.

So if you want to add that while you're in the area I wanted to give you an opportunity to!

**Comment by ewpatton** [comment]
@BeksOmega Let me think about this a bit and get back to you.

File: core/workspace.ts [hunk 1]
```diff
@@ -634,12 +634,12 @@ export class Workspace implements IASTNodeLocation {
       if (!event) continue;
       events.push(event);
     }
+    events = eventUtils.filter(events, redo);
```
Comment (BeksOmega) [inline core/workspace.ts:637]: Is this change still necessary now that we are filtering the undo stack itself?

**Review by BeksOmega — COMMENTED** [review]
Approach LGTM! Not sure if we still need the previous change or not. Once I get clarification on that I'll get this merged =) Thank you again for the fix!

Wrt event groups, they can be interleaved, but we've never handled that for undo and redo (we just stop if we hit a group that's different than the current one, even if the "breaking group" is in the middle of a group). So I see no reason to start handling it now.

File: core/workspace.ts (hunk 1 above)
Comment (ewpatton) [reply core/workspace.ts:637]: It may still be semantically correct to filter events being processed in the undo stack before pushing them to the redo stack. Otherwise, Blockly ends up dropping events from the queue that happen to still be on the stack, which ultimately leads to what happened in the initial issue since there is an event that never gets cleaned up from the stack.

Comment (BeksOmega) [reply core/workspace.ts:637]: So is this just a precautionary thing? My reading is that they're getting filtered before being run + pushed to the undo stack anyway because they're being filtered before they're even pushed to the undo stack.

Comment (ewpatton) [reply core/workspace.ts:637]: Hmm. That is a fair point. I can push an update with only the second commit. If there continue to be occasional issues then we can always reapply it.

**Review by BeksOmega — APPROVED** [review]
